crates/bevy_ecs/src/component.rs

```diff
@@ -138,6 +138,17 @@ impl ComponentInfo {
     }
 }
 
+/// A [`ComponentId`] is an opaque value which uniquely identifies the type of
+/// a [`Component`] within a [`World`](crate::world::World). Each time a new
+/// [`Component`] type is registered within a [`World`](crate::world::World) using
+/// [`World::init_component`](crate::world::World::init_component) or
+/// [`World::init_component_with_descriptor`](crate::world::World::init_component_with_descriptor),
+/// a corresponding [`ComponentId`] is created to track it.
+///
+/// A [`ComponentId`] is tightly coupled to its parent [`World`](crate::world::World).
+/// Attempting to use a [`ComponentId`] from one [`World`](crate::world::World) to access the metadata
+/// of a [`Component`] in a different [`World`](crate::world::World) is undefined behaviour and should
+/// not be attempted.
 #[derive(Debug, Copy, Clone, Hash, Ord, PartialOrd, Eq, PartialEq)]
 pub struct ComponentId(usize);
```

Comment:
I think we should consider talking a bit about the relationship with `TypeId` here too.

Basically, for Rust types, there's a one-to-one mapping between `TypeId` and `ComponentId`. But non-Rust components can be used (well, only kind of so far), and these *also* get their own `ComponentId`.

Reply:
I've added a new section to explain the relationship. I'm not totally happy with the last sentence "Each Rust type ... or other advanced use cases" since it's pretty clunky, but after fiddling with it for a bit I couldn't come up with anything that wasn't quite a bit more verbose. Suggestions appreciated.